### TACOS AND BURRITOS - EXERCISE

**System:**
    
    CPU: i7-7700HQ 2.80GHz
    Cores: 8
    RAM: 16GB
    OS: Linux Mint 19
    
**Software**
    
    Python: 3.7.9
    Apache Spark: 2.4.0

**Dataset**

    Restaurants That Sell Tacos and Burritos (Kaggel)
    https://developer.datafiniti.co/docs/business-data-schema (Support information)

**Tasks:**

    1. Explore and clean the column 'cuisines'.
    2. Create a ranking of most popular cuisines. (Hint: MapReduce - Flatmap)

#### SOLUTION (PANDAS)

**Load packages and data**

In [266]:
import pandas as pd

df = pd.read_csv("/home/agomez/MEGA/Cursos/Linkedin-Spark/tacos_and_burritos.csv")

**Dataframe summary**

In [179]:
len(df)

77260

In [180]:
df.describe() # Numerical columns

latitude     longitude  menus.amountMax  menus.amountMin  \
count  55636.000000  55636.000000     40260.000000     40260.000000   
mean      36.694846    -98.713309         8.754767         8.611666   
std        4.835124     18.245857        18.408596        15.708317   
min      -31.986438   -159.492690         0.000000         0.000000   
25%       33.668355   -117.647150         5.000000         5.000000   
50%       36.047195    -96.682320         7.950000         7.950000   
75%       40.588380    -82.679930        10.500000        10.390000   
max       61.219460    115.903696      1990.000000      1290.000000   

       priceRangeMin  priceRangeMax  Unnamed: 26  Unnamed: 27  Unnamed: 28  \
count   48159.000000   48159.000000          0.0          0.0          0.0   
mean       13.734837      33.722627          NaN          NaN          NaN   
std        14.704344      10.395812          NaN          NaN          NaN   
min         0.000000       5.000000          NaN          NaN          NaN   
25%         0.000000      25.000000          NaN          NaN          NaN   
50%        25.000000      40.000000          NaN          NaN          NaN   
75%        25.000000      40.000000          NaN          NaN          NaN   
max       378.000000     755.000000          NaN          NaN          NaN   

       Unnamed: 29  ...  Unnamed: 226  Unnamed: 227  Unnamed: 228  \
count          0.0  ...           0.0           0.0           0.0   
mean           NaN  ...           NaN           NaN           NaN   
std            NaN  ...           NaN           NaN           NaN   
min            NaN  ...           NaN           NaN           NaN   
25%            NaN  ...           NaN           NaN           NaN   
50%            NaN  ...           NaN           NaN           NaN   
75%            NaN  ...           NaN           NaN           NaN   
max            NaN  ...           NaN           NaN           NaN   

       Unnamed: 229  Unnamed: 230  Unnamed: 231  Unnamed: 232  Unnamed: 233  \
count           0.0           0.0           0.0           0.0           0.0   
mean            NaN           NaN           NaN           NaN           NaN   
std             NaN           NaN           NaN           NaN           NaN   
min             NaN           NaN           NaN           NaN           NaN   
25%             NaN           NaN           NaN           NaN           NaN   
50%             NaN           NaN           NaN           NaN           NaN   
75%             NaN           NaN           NaN           NaN           NaN   
max             NaN           NaN           NaN           NaN           NaN   

       Unnamed: 234  Unnamed: 235  
count           0.0           0.0  
mean            NaN           NaN  
std             NaN           NaN  
min             NaN           NaN  
25%             NaN           NaN  
50%             NaN           NaN  
75%             NaN           NaN  
max             NaN           NaN  

[8 rows x 216 columns]

In [133]:
df.describe(include='O') # Categorical columns

id             address          categories  \
count                  77260               77260               77260   
unique                 19439               18618                8488   
top     AVwch4pVByjofQCxfPyl  3000 N Dinuba Blvd  Mexican Restaurant   
freq                      68                  69               15094   

             city country cuisines             dateAdded  \
count       77260   77260    38384                 77260   
unique       3601       1     1588                 17965   
top     San Diego      US  Mexican  2015-10-23T02:21:55Z   
freq         1924   77260     6280                    79   

                 dateUpdated                                            keys  \
count                  77260                                           77260   
unique                 17675                                           19438   
top     2015-11-02T15:47:29Z  us/or/portland/10090sewashingtonst/-1322134216   
freq                      68                                              68   

                                              menuPageURL menus.category  \
count                                               10054           3729   
unique                                               2600            962   
top     http://www.yellowpages.com/long-beach-ca/mip/e...          Tacos   
freq                                                   53            291   

       menus.currency        menus.dateSeen  \
count           40511                 77260   
unique              2                 18453   
top               USD  2016-04-24T00:00:00Z   
freq            40495                   821   

                                        menus.description         menus.name  \
count                                               29323              77260   
unique                                              15730              15737   
top     Crunchy corn tortilla, spicy beef, american ch...  Breakfast Burrito   
freq                                                 1087               3188   

                       name postalCode priceRangeCurrency province  \
count                 77257      76837              48159    77257   
unique                10526       7463                  2     1097   
top     Chili's Grill & Bar      92110                USD       CA   
freq                   8296        240              48158    23664   

                 websites  
count               55989  
unique               9238  
top     http://chilis.com  
freq                 2923

The dataframe contains a total of 236 columns. There are 216 numeric columns and 20 categorical columns. The total number of rows is 77.260. The first 26 columns are named while the rest are called 'Unnamed: NUM'.

In [267]:
df = df.iloc[:,0:26] # Select range of columns

**TASK 1**

In [182]:
print("Before clean, length dataframe:",len(df))

Before clean, length dataframe: 77260


In [268]:
# Clean NAs or NaN from the column cuisine

df = df.dropna(subset = ["cuisines"])

The data is not duplicated so we do not need to clean it after removing the NAs.

In [184]:
print("After clean, length dataframe:",len(df))

After clean, length dataframe: 38384


In [185]:
df["cuisines"].head(20)

0      Restaurant Delivery Service
1      Restaurant Delivery Service
4                          Mexican
5                          Mexican
6                          Mexican
7                          Mexican
8                          Mexican
9                          Mexican
10                         Mexican
11                         Mexican
12                         Mexican
13                         Mexican
14                         Mexican
15                    Coffee Shops
16                    Coffee Shops
17                    Coffee Shops
18         Mexican, Latin American
19         Mexican, Latin American
21    American,Seafood Restaurants
22    American,Seafood Restaurants
Name: cuisines, dtype: object

There are no NaN values but we still have an issue with the data, some rowd contains values divided by comma. Example, 'Mexican, Latin American'. We need to remove that issues.

In [300]:
# Match name and cuisine after split by comma cuisine and finally select column 'index'

df2 = pd.concat([pd.Series(row["name"], row["cuisines"].split(',')) for _, row in df.iterrows()]).reset_index()
df2.rename(columns={'index':'cuisines',0:'name'}, inplace=True)

In [272]:
# Check result

# df2["cuisines"].unique()

We have corrected the comma separated rows but we can see that there are slash (between spaces) separated rows. For example, "Bar / Lounge / Bottle Service". We should not worry about it since is the same definition. However, we can divide them using the following code:

In [126]:
# Match name and cuisine after split by slash between spaces ( / ) cuisine 
# and finally select column 'index'

# df2 = pd.concat([pd.Series(row["name"], row["cuisines"].split(' / ')) for _, row in temp.iterrows()]).reset_index()
# df2.rename(columns={'index':'cuisines',0:'name'}, inplace=True)

In [273]:
# Check result

# df2["cuisines"].unique()

Finally, the last issue we have is that there are cuisines that for us are equal but for the count seem different. For example, ' Mexican' and 'Mexican' (space before word). Also, 'Mexican' and 'Mexican Restaurants' could be an issue because we know both are Mexican cuisines but they are defined different. We will only correct the first issues to make it simple.

Also, before selecting the column cuisine we need to get a unique hit per food place (name) so we avoid repeated entries.

In [302]:
#Remove duplicates
df2 = df2.drop_duplicates()

In [307]:
# Remove spaces from values
df2 = df2["cuisines"].apply(lambda x: x.strip(" "))

In [308]:
# Check that it worked
# for i in df2["cuisines"].apply(lambda x: x.strip(" ")).unique():
#    if 'Mexican' in i:
#        print(i)

In [309]:
df2

0        Restaurant Delivery Service
2                            Mexican
13                      Coffee Shops
16                           Mexican
17                    Latin American
                    ...             
55730                        Mexican
55737                        Mexican
55738                        Spanish
55743                            Pub
55754                        Mexican
Name: cuisines, Length: 9872, dtype: object

Our data is clean now and we can go to the second task. We have a total of 9.872 entries (it should be less if we account for equal types with different labels).

**TASK 2**

Using the support information, for the second task, we are interested in the columns:

    'name'
    'cuisines'

In [311]:
# Count the number of times a  cuisine appears.

result = df2.value_counts()

In [312]:
result[0:10]

American                      1130
Mexican                        931
Mexican Restaurants            438
Contemporary American          426
Take Out Restaurants           409
Seafood                        367
American Restaurants           321
Restaurants                    308
Pizza                          271
Latin American Restaurants     241
Name: cuisines, dtype: int64

The top 10 ranking of best cuisines in the database shows that the most populat cuisines is _American_ with 1130 food places. It is followed by _Mexican_ and _Mexican Restaurants_.

However, if we sum the 2nd and 3rd option (which are equal in a sense) and the 1st and 8th option, we end up with:

    American 1130 + 321 = 1451
    Mexican  931  + 438 = 1369
    
The first place is still for _American_ followed by _Mexican_.

##### SOLUTION (PYSPARK)

In [1]:
from pyspark import SparkConf, SparkContext
from pyspark.sql.types import StringType
from pyspark import SQLContext
from pyspark.sql.functions import col

conf = SparkConf().setMaster("local").setAppName("Local-Spark")
sc = SparkContext(conf = conf)
sqlContext = SQLContext(sc)

In [101]:
# Read data

dfspark = sqlContext.read.format("csv").option("header","true").option("inferSchema","true").load("/home/agomez/MEGA/Cursos/Linkedin-Spark/tacos_and_burritos.csv")

In [102]:
# Select columns of interest (name & cousines)

dfspark = dfspark.select("name","cuisines")

In [103]:
# Show data

dfspark.take(5)

[Row(name="Gambino's Pizza", cuisines='Restaurant Delivery Service'),
 Row(name="Gambino's Pizza", cuisines='Restaurant Delivery Service'),
 Row(name='Diablo Grande Golf & Country Club', cuisines=None),
 Row(name="Mcdonald's", cuisines=None),
 Row(name='Baja Fresh Mexican Grill', cuisines='Mexican')]

In [104]:
# Clean NaN

dfspark = dfspark.na.drop(subset = ["name","cuisines"])

In [105]:
# Remove duplicates, after ordering them by name, for same restaurant/food place

dfspark = dfspark.orderBy("name").dropDuplicates(subset = ["name"])

In [106]:
# Show data

dfspark.take(5)

[Row(name='#9 Steakhouse', cuisines='Steak Houses'),
 Row(name='$5 Pizza', cuisines='Pizza'),
 Row(name="'Tis Corner Cafe", cuisines='Bagels'),
 Row(name='024 Grille', cuisines='Contemporary American, American, Wine Bar'),
 Row(name='100% De Agave', cuisines='Contemporary Mexican, Bar / Lounge / Bottle Service,Mexican')]

After cleaning, we have the name one time and the cuisine type of the food place. Now, we can divide the data given the delimiter. In this case, we can stop using the column 'name'.

In [189]:
# Parallelize 

pl_df = sc.parallelize(dfspark.select("cuisines").rdd.collect())

In [190]:
def split_delimiters(row):
    row = row[0]
    row = row.replace(" ","")
    row = row.split(",")
    return row

In [191]:
results = pl_df.flatMap(split_delimiters)

In [192]:
results.collect()[0:10]

['SteakHouses',
 'Pizza',
 'Bagels',
 'ContemporaryAmerican',
 'American',
 'WineBar',
 'ContemporaryMexican',
 'Bar/Lounge/BottleService',
 'Mexican',
 'Casinos']

In [193]:
# Add a 1 at the right of each value

results = results.map(lambda cousin: (cuisin,1))

In [194]:
# Some all the 1's in the right of the values with same key

results = results.reduceByKey(lambda cuisin,num: cuisin+num)

In [196]:
results.sortBy(lambda tupple: tupple[1], ascending=False).collect()

[('American', 1071),
 ('Mexican', 819),
 ('ContemporaryAmerican', 422),
 ('MexicanRestaurants', 388),
 ('Seafood', 360),
 ('TakeOutRestaurants', 351),
 ('AmericanRestaurants', 292),
 ('Pizza', 251),
 ('Restaurants', 245),
 ('LatinAmericanRestaurants', 216),
 ('CoffeeShops', 173),
 ('Bars', 151),
 ('Italian', 151),
 ('ComfortFood', 139),
 ('TraditionalAmerican', 135),
 ('Californian', 107),
 ('Steakhouse', 106),
 ('Steak', 102),
 ('Caterers', 96),
 ('International', 82),
 ('Bar/Lounge/BottleService', 81),
 ('GastroPub', 80),
 ('Japanese', 74),
 ('LatinAmerican', 73),
 ('Mediterranean', 73),
 ('Burgers', 70),
 ('Asian', 70),
 ('Tapas/SmallPlates', 67),
 ('Sushi', 64),
 ('FastFoodRestaurants', 64),
 ('Mexican/Southwestern', 59),
 ('Fusion/Eclectic', 58),
 ('FamilyStyleRestaurants', 58),
 ('ContemporaryMexican', 52),
 ('SeafoodRestaurants', 49),
 ('Vegetarian', 48),
 ('Global', 48),
 ('NewAmerican', 48),
 ('Barbecue', 48),
 ('Spanish', 46),
 ('Organic', 46),
 ('SushiBars', 42),
 ('ItalianR

**Or**

In [223]:
# As before
results = pl_df.flatMap(split_delimiters)

In [224]:
# Use function countByValue() which returns a dictionary
results = results.countByValue()

In [213]:
# Sort dictionary (reverse) and show top 10

sorted(results, key=results.get, reverse=True)[0:10]

['American',
 'Mexican',
 'ContemporaryAmerican',
 'MexicanRestaurants',
 'Seafood',
 'TakeOutRestaurants',
 'AmericanRestaurants',
 'Pizza',
 'Restaurants',
 'LatinAmericanRestaurants']

In [225]:
# Finally, print keys and values

top10 = sorted(results, key=results.get, reverse=True)[0:10]

for value in top10:
    print(value, ":", results[value])

American : 1071
Mexican : 819
ContemporaryAmerican : 422
MexicanRestaurants : 388
Seafood : 360
TakeOutRestaurants : 351
AmericanRestaurants : 292
Pizza : 251
Restaurants : 245
LatinAmericanRestaurants : 216
